### DISCUSSION TITLE
Access multiple data sets by name

### DISCUSSION THREAD

**Discussion by psheehymn** [gapitio/gapit-htmlgraphics-panel#166] [Q&A]
I have a panel with numerous queries, each of which returns a single data set consisting of path names within our portal and various single value metrics (page hit count, percentile load time, HTTP error count, etc.). I access each query by `refId` and retrieve the path names and values similar to the following (this is simplified from the real code for illustration):
```
let HitsQ = data.series.find(series => series.refId == "Hits");
for (let i = 0; i < HitsQ.fields[0].values.length; i++) {
  hitPathName[i] = HitsQ.fields[0].values.get(i);
  hitPathValue[i] = HitsQ .fields[1].values.get(i);
}
```
I can then align all the data from the various queries by the path names into a nice table format.

I now have to add a different type of query that returns _only_ timeseries data sets, one for each path name. In a stand-alone panel, I can get this to also work out by cycling thru all the data sets to retrieve the path names and sum the timeseries values as follows:
```
for (let i = 0; i < data.series.length; i++) {
  mHitsPathName[i] = data.series[i].fields[1].name;
  mHitsPathValue[i] = data.series[i].fields[1].state.calcs.sum;
}
```
Referencing the `refId` of all the data sets (i.e. `data.series[i].fields[1].refId`) show they all have the same name (the `refId` of the original query).

In order to integrate this query into the first panel, I need to access the query by name. If I try to access the query by `refId` as I do for the other queries, and then ask for the path name and a count:
```
let mHitsQ = data.series.find(series => series.refId == "mHits");
console.log("mHits name="+mHitsQ.fields[1].name+" mHits count="+mHitsQ.length);
```
it gives me path name and the number of entries in the first dataset. Any attempt to use array notation with mHitsQ to get at the other data sets generate an error. 

I know I can loop through the entire array of data sets in `data.series` and find the ones related to this query's `refId`, but is there a more efficient access by name syntax I am missing that will return an array of _only_ those data sets associated with the `refID`?

**Comment by ZuperZee** [comment]
> Referencing the refId of all the data sets (i.e. data.series[i].fields[1].refId) show they all have the same name (the refId of the original query).

Usually the refId is located at `data.series[i].refId`.

> In order to integrate this query into the first panel, I need to access the query by name. If I try to access the query by refId as I do for the other queries, and then ask for the path name and a count:
...
it gives me path name and the number of entries in the first dataset. Any attempt to use array notation with mHitsQ to get at the other data sets generate an error.

To me, it sounds like redId "mHits" doesn't exist.
Your series object structure might be different from the other query type.

> I know I can loop through the entire array of data sets in data.series and find the ones related to this query's refId, but is there a more efficient access by name syntax I am missing that will return an array of only those data sets associated with the refID?

Not really. If you want to get the series with the refId from the query, that's the easiest.

The above might not be of much help, but if you can create a minimal example of your data.series object, I can guide you a bit more. **Don't add any sensitive information. Try to replace your values with some random stuff or create it from scratch.**

Example data.series object:
```js
[
    {
        "refId": "A",
        "fields": [
            {
                "name": "time",
                "type": "time",
                "values": [
                    1698412519789
                ],
                "state": {
                    "seriesIndex": 0
                }
            },
            {
                "name": "Something",
                "type": "number",
                "values": [
                    1
                ],
                "state": {
                    "seriesIndex": 0,
                    "range": {
                        "min": 1,
                        "max": 1,
                        "delta": 0
                    },
                    "calcs": {
                        "sum": 1,
                        "max": 1,
                        "min": 1,
                        "logmin": 1,
                        "mean": 1,
                        "last": 1,
                        "first": 1,
                        "lastNotNull": 1,
                        "firstNotNull": 1,
                        "count": 1,
                        "nonNullCount": 1,
                        "allIsNull": false,
                        "allIsZero": false,
                        "range": 0,
                        "diff": 0,
                        "delta": 0,
                        "step": null,
                        "diffperc": 0,
                        "previousDeltaUp": true
                    },
                    "displayName": "Something",
                    "multipleFrames": true
                }
            }
        ],
        "length": 1
    },
    {
        "refId": "B",
        "fields": [
            {
                "name": "time",
                "type": "time",
                "values": [
                    1698412519789
                ],
                "state": {
                    "seriesIndex": 1
                }
            },
            {
                "name": "Something",
                "type": "number",
                "values": [
                    1
                ],
                "state": {
                    "calcs": {
                        "sum": 1,
                        "max": 1,
                        "min": 1,
                        "logmin": 1,
                        "mean": 1,
                        "last": 1,
                        "first": 1,
                        "lastNotNull": 1,
                        "firstNotNull": 1,
                        "count": 1,
                        "nonNullCount": 1,
                        "allIsNull": false,
                        "allIsZero": false,
                        "range": 0,
                        "diff": 0,
                        "delta": 0,
                        "step": null,
                        "diffperc": 0,
                        "previousDeltaUp": true
                    },
                    "seriesIndex": 1,
                    "range": {
                        "min": 1,
                        "max": 1,
                        "delta": 0
                    },
                    "displayName": "Something",
                    "multipleFrames": true
                }
            }
        ],
        "length": 1
    }
]
```

**Reply by psheehymn** [reply]
>Usually the refId is located at data.series[i].refId.

You're correct, I was referencing from memory and messed up the syntax.

Looping through the array of data sets was straight forward enough, I went ahead with that solution and it works well. It's not worth spending a lot of additional time trying to come up with a more "elegant" solution if you don't already know of one (and especially if there likely isn't one). You can mark this one complete.

BTW, thank you for the sample data.series; I am saving that for future reference!
